### ISSUE TITLE
Guidelines for Examples

### ISSUE THREAD

**Issue by Marvin182** [google/flax#231]
Following some discussion I wrote down a set of guidelines for Flax examples inside the google/flax repository. It will take some work to update the current set of examples but I believe that the examples will be even better afterwards and easier to maintain.

## Guidelines
- Each example should have at least 2 owners . Creating examples cost time and we want to recognize this. We also would like to have a point of contact in case the example needs to be updated to a new API.
- Every example should have a `README.md` which specifies at least:
  - Command for running the example with a link to http://tensorboard.dev for a successful run. Optionally one can also include some output logged on the command line.
  - Link to related references (paper for the model or a great blogpost).
- Keep the example focused.
  - Each example should support a single model on a single dataset.
  - Keep the number of configurable hyper-parameters basic (<10) and use `absl.flags`. We do not want to overwhelm readers.
- Default hyperparameter configuration should be provided as ConfigDict in `configs/default.py`. An additional `configs/test.py` for a CPU friendly unit test case are ok.
- To make it easier to test, maintain and reuse code structure the code as follows:
  - `main.py` contains the flags and calls a method from `train.py` to run the training loop. This should be the only file defining flags! This can be almost identical for all examples, please copy from `linen_examples/wmt/main.py`. If possible stick with the 2 flags (`--workdir` (not `model_dir`) and `--config`).
  - `train.py` contains classes and methods for training and evaluating the model.
  - `train_test.py` for test cases of the training code. At a minimum the test should run a single training step but more fine grained unit tests are a bonus. You can use `tfds.testing.mock_data` to avoid real data from disk.
  - Additional files for more complex architectures or `input_pipeline.py`.
- Only use public datasets available in [TensorFlow Datasets](https://www.tensorflow.org/datasets). TensorFlow Datasets standardizes how datasets downloaded and prepared. Synthetic data is also allowed.
- Follow the [Google Python Style Guide](http://google.github.io/styleguide/pyguide.html) (no relative imports, use type annotation)
- `requirements.txt` if running the example requires additional Python packages that are not installed by Flax or it's dependencies.
- Use the `clu` package where appropriate. This helps us to de-duplicate common functionality and make the code agnostic to the platform. We would like to see examples that work equally well on Google's internal systems as on Cloud or locally.
- 

## Future Work
Examples should also be covered by regression tests, see #144.

We should also add general instructions for running on examples (on GPU and TPU) using Google Cloud.

**Comment by danielsuo** [comment]
Thanks for writing this up in much more detail! For sake of #230, two things:

1. The testing code should not use any `absl.flags` or call code that uses flags (i.e., in the code structure you propose, only call from `{example}_lib.py`).
2. Not sure what to do about no relative imports. `pytest` gets confused if we call from the root `flax` directory without them and `examples` is not a package. Suggestions?

**Comment by mohitreddy1996** [comment]
Thanks a lot and +1 to the proposal Marvin!

As part of #176, I did face issues with using tfds.testing.mock_data and required refactoring the code in train.py often (given the how different creating model and optimizers are in each example). I believe this will definitely make the tests more readable and easier to implement.

**Comment by danielsuo** [comment]
Disregard my earlier comment; I think I figured it out. Thanks for the push!

**Comment by avital** [comment]
@BertrandRdp was also looking into a (mostly) orthogonal task which was to clarify guidance around which examples we have in flax/examples vs how we would link to external repositories, and how we would help people discover what models are desired by the community.

**Comment by avital** [comment]
One comment:

> Follow the Google Python Style Guide (no relative imports, use type annotation)
1. No relative imports: Our examples are not modules so we use `import foo` to import other files in that directory. Not sure how that relates to the "no relative imports" directive

2. Use type annotation: I don't know if this is a good idea for an example model training codebase. What's the upside?

**Comment by avital** [comment]
@Marvin182 would be great to see what one example ported to your proposed format would look like, then we can split up the examples and have different people port all of the examples.

**Comment by avital** [comment]
@mohitreddy1996 has agreed to get the ball rolling on this by porting MNIST and/or ImageNet. We'll review that and then if a quorum of us like it, we can split up the work for porting the other examples.

That should also help unblock #176

**Comment by mohitreddy1996** [comment]
@Marvin182 @avital 

Marvin do you think 'example_' prefix in the file name is required. 

IMO this could be redundant given the files already exist in a dir 'example/'. WDYT? I have no strong opinion, just throwing a thought.

**Comment by avital** [comment]
> Marvin do you think 'example_' prefix in the file name is required.
One bizarre upside is that it would simplify the code we use to run tests, which means it's easier to run tests within an IDE rather than calling our custom script, see #230

**Comment by marcvanzee** [comment]
Nice proposal! One nit:

> Only use public datasets available in TensorFlow Datasets. TensorFlow Datasets standardizes how datasets downloaded and prepared.

I guess using simple synthetic datasets should also be allowed, which now happens in the seq2seq example: https://github.com/google/flax/blob/master/examples/seq2seq/train.py

**Comment by andsteing** [comment]
(Will be updated by edits)

We started rewriting the examples one by one:

- [ ] `mnist` #479 
- [ ] `imagenet` #480 

Some small updates on what @Marvin182 described in his original proposal:

- Every example should contain a Jupyter notebook / Colab for illustration purposes and interactive development.
- Hyper parameter configuration : Use [`ml_collections.ConfigDict`](https://github.com/google/ml_collections#configdict) and [`ml_collections.config_flags`](https://github.com/google/ml_collections#config-flags) (e.g. #492).
- Code structure
  - Suggested additional model file names: `models.py`, `input_pipeline.py`
- Regression tests : Examples should have `{example}_fake_data_benchmark.py` that run quickly and `{example}_benchmark.py` with real data that are run at fixed intervals.

Note: Since we're currently also refactoring the examples for the new [Linen API](https://github.com/google/flax/tree/master/flax/linen), examples will be updated in the [`linen_examples`](https://github.com/google/flax/tree/master/linen_examples) directory that will eventually replace the current [`examples`](https://github.com/google/flax/tree/master/examples) directory.

**Comment by mortner31** [comment]
Dear all, 

Thanks for your awesome work. I have been digging intensively around the linen examples and maybe my feedback regarding what is not clear or not easy to cope with maybe of some use. Overall the examples are great, some minor points of discussion here. 

**About Models, JIT, module instanciation and static_argnums**  

In the former examples  - without  linen -  the use of `nn.Model` allowed to hide the instanciation complexity inside jitted functions.  Basically, the `Model` object would port a callable creating the module instance. The `nn.Model` was made jittable  as such : 

```python
@struct.dataclass
class Model:
  """A Model contains the model paramaters, state and definition."""

  module: Module = struct.field(pytree_node=False)
  params: Any 
```

As a consequence , the  `train_step` and `eval_step` were directly made jaxable by use of `Model` instances while calling the functions.  In the linen api example, model creation inside the Jitted functions can be done very differently.  

Use of direct call 
```python
#from mnist example
@jax.jit
def train_step(optimizer, batch):
  """Train for a single step."""
  def loss_fn(params):
    logits = CNN().apply({'params': params}, batch['image'])
    loss = cross_entropy_loss(logits, batch['label'])
    return loss, logits
```
... by use of a special `config` 
```python 
# from wmt example
# call a jitted initialization function to get the initial parameter tree  
@jax.jit
def initialize_variables(rng):
   return models.Transformer(eval_config).init(
        rng,
        jnp.ones(input_shape, jnp.float32),
        jnp.ones(target_shape, jnp.float32))
```
where the config is declared as a jaxable struct using flax specific dataclass decorator
```python
@struct.dataclass
class TransformerConfig:
  """Global hyperparameters used to minimize obnoxious kwarg plumbing."""
  vocab_size: int
  output_vocab_size: int
  # [...] 
```
of even by use of globals 
```python
# from Seq2Seq example
def model(teacher_force=True):
  return Seq2seq(eos_id=CTABLE.eos_id, teacher_force=teacher_force,
                 hidden_size=FLAGS.hidden_size)

```

I think all these different methods make it unclear that the `train_step` and `eval_step` are functions that should obey the jax principles.  As a consequence of this lack of clarity, I typically ended up with a different version for my teammates,  that exhibits the fact that the module constructor should be a `static_argnum` because it will be computed only once due to jax mechanism.

```python
# jax will recompile if module_constructor is changed. Since it is a callable, 
# comparison will be done based of object id. 
@jax.partial(jax.jit, static_argnums=3)
def eval_step(state, batch_image, batch_label, module_constructor : Callable):
    params = state.all_params
    module = module_constructor(is_training=False)
    logits = module.apply(params, batch_image)
    return compute_metrics(logits, batch_label)
```

**Few example with stochastic  (rng keys)**

It is a small comment, but I found two version.

One where dropout keys are updated inside the train_step
```python
# from wmt
    optimizer, metrics, dropout_rngs = p_train_step( optimizer, batch, dropout_rng=dropout_rngs)
```

And in the `vae` example where the keys are splitted outside the train step
```python
# from vae example  
for epoch in range(FLAGS.num_epochs):
    for _ in range(steps_per_epoch):
      batch = next(train_ds)
      rng, key = random.split(rng)
      optimizer = train_step(optimizer, batch, key)

```
Both way of using `rngs` are fine, I think, but it is hidden in the "larger examples" and I have the feeling that some users may forgot to update keys at each iteration. 

 
**Jit and pmap**

Another (small thing) is that the `imagenet` example rrelies on `pmap` . It therefore hides the fact the `train_step` and `eval_step` are `jitted`. It may be worth adding a comment, just  in order to go from 

```python 
@jax.jit
def eval_step(state, batch):

```
to 
```python 
# to be jaxed by jax.pmap
def eval_step(state, batch):
```
instead of 
```python 
def eval_step(state, batch):
```

**Proposition for easy understanding**

I understand the logic of presenting many differing examples covering all use cases. However, I think it might be worth proposing several forked mnist examples  in a family of examples. 

* A basic mnist example (as proposed today) 
* An mnist example with model structure parametrization (using config or anything) 
* An mnist example with batch norm version  (to illustrate the `batch_stats` usage)
* An mnist example with dropout (to illustrate `rngs` usage) 
* An mnist example with pmap (to illustrate the distributed gradient usage) 
* An mnist example with checkpointing

Such a stack of mnist example would allow to work by diff and understand very quickly the different elementary concepts. The examples as a whole cover all cases, but each of the example differ slightly so it is sometimes difficult to separate each case. I understand that maintenance of separate "only slightly differing" example is a pain, but it could really be helpful.

**Comment by andsteing** [comment]
Hi Mathias

Thanks for your detailed suggestions. Some answers/suggestions:

As for the `Model` abstraction : we don't currently use this anymore in Linen, but @jheek is currently reworking the imagenet example in #506 to pass around the `module.apply()` function instead of re-creating the module on the fly:

https://github.com/google/flax/blob/b928699048467eccc7b6b3e3633b84e19d36f7a2/linen_examples/imagenet/imagenet_lib.py#L307-L310

Note that as per this issue we will update all examples to use `ml_collections.ConfigDict` to pass hyper parameters and this `config` object could always be used to recreate a new module (though as of today the object is not JIT-table so you would rather use above pattern).

About `rngs` : Yes it would probably make sense to use the new Linen `rng` API in the `vae` example as well, feel free to create a PR for that one! In general I think it's nice having uniformity in examples, but sometimes it can also be interesting to show different ways of achieving the same goal. Same for the "Jit and pmap".

Regarding "Proposition for easy understanding" : That's an excellent idea! Instead of having multiple forks of the `mnist` example we rather create HOWTOs on top of the one checked in example and make sure they're updated in the same PR if needed. We already have two of the diffs that you suggest:

- https://flax.readthedocs.io/en/latest/howtos/distributed-training.html
- https://flax.readthedocs.io/en/latest/howtos/checkpointing.html

Feel free to create issues / PRs for more HOWTOs : https://github.com/google/flax/issues?q=is%3Aopen+is%3Aissue+howto

**Comment by andsteing** [comment]
### Conventions

This comment collects some conventions that all examples should adhere by for better inter-example consistency.

Note that all these suggestions are open to discussion, and that this comment is expected to be modified and updated as the list of conventions grows and improves.

- For models with dropout, the parameter `train` should be specified directly to the `__call__()` method and it should not have a default.

**Comment by Marvin182** [comment]
I updated the proposals with some minor changes which will help us that examples work on Google's internal and external Cloud offerings.

**Comment by marcvanzee** [comment]
@andsteing do you think this issue should say open, and if so why?
If not, shall we move the "Example Guidelines" into another doc and close this one?

**Comment by andsteing** [comment]
Good idea. Let me summarize this into a standalone RST file that we can reference and close the issue here.

**Comment by andsteing** [comment]
Closing this issue with #1841

**Comment by 1kaiser** [comment]
can the dataset loading mechanism be focused with more concise custom datasets too , such as dataset creation in tfds format quickly <<< so that it saves time >>>
